### PR TITLE
Add BasicAuth security in OpenAPI specs

### PR DESCRIPTION
**Pull request by nitinmohan87**
Goa implicitly sets Authorization header when an endpoint is secured with BasicAuth. But OpenAPI specs don't show the endpoint as secured at all. Hence we need to explicitly add the Authorization header in the OpenAPI spec when using BasicAuth.

### CODE REVIEW

File: expr/http_endpoint.go [hunk 1]
```diff
@@ -660,44 +660,67 @@ func (e *HTTPEndpointExpr) validateParams() *eval.ValidationErrors {
...
-		if IsObject(nat.Attribute.Type) {
-			verr.Add(e, "header %s cannot be an object, header type must be primitive or array", nat.Name)
-		} else if arr := AsArray(nat.Attribute.Type); arr != nil {
-			if !IsPrimitive(arr.ElemType.Type) {
-				verr.Add(e, "elements of array header %s must be primitive", nat.Name)
+
+	// We have to figure out the actual type for the params because the actual
+	// type is initialized only during the finalize phase. In the validation
+	// phase, all param types are string type by default unless specified
+	// expliclty.
```
Comment (raphael) [inline expr/http_endpoint.go]: typo

File: expr/http_endpoint.go [hunk 2]
```diff
@@ -660,44 +660,67 @@ func (e *HTTPEndpointExpr) validateParams() *eval.ValidationErrors {
...
+	init := func(ma *MappedAttributeExpr) *MappedAttributeExpr {
+		ma = DupMappedAtt(ma)
```
Comment (raphael) [inline expr/http_endpoint.go]: Can this be encapsulated in a private function that is used by both the validation and finalization steps?

File: expr/http_endpoint.go [hunk 3]
```diff
@@ -660,44 +660,67 @@ func (e *HTTPEndpointExpr) validateParams() *eval.ValidationErrors {
...
+
+	headers := init(e.Headers)
+	WalkMappedAttr(headers, func(name, _ string, a *AttributeExpr) error {
+		switch {
+		case IsObject(a.Type):
+			verr.Add(e, "header %s cannot be an object, header type must be primitive or array", name)
+		case IsArray(a.Type):
+			arr := AsArray(a.Type)
+			if !IsPrimitive(arr.ElemType.Type) {
+				verr.Add(e, "elements of array header %s must be primitive", name)
```
Comment (raphael) [inline expr/http_endpoint.go]: `%q` might be better than `%s`?

File: expr/http_endpoint.go [hunk 4]
```diff
@@ -660,44 +660,67 @@ func (e *HTTPEndpointExpr) validateParams() *eval.ValidationErrors {
...
-			}
-		case *Map:
-			if len(*headers) > 0 {
-				verr.Add(e, "Payload type is map but HTTP endpoint defines headers. Map payloads can only be decoded from HTTP request bodies or query strings.")
+
+	headers := init(e.Headers)
+	WalkMappedAttr(headers, func(name, _ string, a *AttributeExpr) error {
+		switch {
+		case IsObject(a.Type):
+			verr.Add(e, "header %s cannot be an object, header type must be primitive or array", name)
```
Comment (raphael) [inline expr/http_endpoint.go]: `%q` might be better than `%s`?

File: expr/http_endpoint.go [hunk 5]
```diff
@@ -660,44 +660,67 @@ func (e *HTTPEndpointExpr) validateParams() *eval.ValidationErrors {
...
+		case IsArray(a.Type):
+			arr := AsArray(a.Type)
+			if !IsPrimitive(arr.ElemType.Type) {
+				verr.Add(e, "elements of array header %s must be primitive", name)
 			}
+		default:
+			ctx := fmt.Sprintf("header %s", name)
```
Comment (raphael) [inline expr/http_endpoint.go]: `%q` might be better than `%s`?

File: http/codegen/openapi/openapi_v2_builder.go [hunk 6]
```diff
@@ -158,8 +158,8 @@ func securitySpecFromExpr(root *expr.RootExpr) map[string]*SecurityDefinition {
 			sd.Type = "basic"
 		case expr.APIKeyKind:
 			sd.Type = "apiKey"
-			sd.In = s.In
-			sd.Name = s.Name
+			sd.In = "header"
```
Comment (raphael) [inline http/codegen/openapi/openapi_v2_builder.go]: Why are we hardcoding here? API keys can be set in parameters as well for example (maybe we need a test for that).

Comment (nitinmohan87) [reply http/codegen/openapi/openapi_v2_builder.go]: This defines the top-level `securityDefinitions` object. It is possible that the there are two endpoints secured by API key where one passes the key in the header and the other passes in the param, correct?

File: expr/http_endpoint.go [hunk 7]
```diff
@@ -638,44 +628,51 @@ func (e *HTTPEndpointExpr) validateParams() *eval.ValidationErrors {
...
 	verr := new(eval.ValidationErrors)
-	for _, nat := range *headers {
-		if IsObject(nat.Attribute.Type) {
-			verr.Add(e, "header %s cannot be an object, header type must be primitive or array", nat.Name)
-		} else if arr := AsArray(nat.Attribute.Type); arr != nil {
+
+	// We have to figure out the actual type for the headers because the actual
+	// type is initialized only during the finalize phase. In the validation
+	// phase, all param types are string type by default unless specified
+	// explicity.
```
Comment (raphael) [inline expr/http_endpoint.go]: Typo

**Pull request merged**